### ISSUE TITLE
Inexact operators for prerelease version constraints

### ISSUE THREAD

**Issue by Jellayy** [opentofu/opentofu#1651]
### OpenTofu Version

```shell
❯ tofu version
OpenTofu v1.7.1
on darwin_arm64
```

### Use Cases

I help to maintain modules that deploy across a large number of workspaces and want to integrate a tiered deployment strategy. That involves releasing alpha, beta, and release candidate releases of these modules alongside mainline releases.

I have a groups of workspaces that I'd like to pull in these different prerelease channels. Ideally, that would be accomplished by setting these workspaces to take in the module with something similar to `>= 1.0.0-beta.1` or `~> 1.0-rc.1`. However, [prerelease versions can currently only be selected by an exact version constraint](https://opentofu.org/docs/language/expressions/version-constraints/#version-constraint-behavior).

### Attempted Solutions

On a module that has the versions: `1.0.1`, `1.1.0-beta.1`, `1.1.0-beta.2`, `1.1.0-beta.3`, `1.1.0-rc.1` available, setting the version constraint to an inexact operator such as `>= 1.1.0-beta.1` results in the following:
```
There is no available version of module "..." (main.tf:1) which matches the given version constraint. The newest available version is 1.0.1.
```

However, setting the version constraint to an exact operator such as `1.1.0-beta.1` pulls down the module version properly.

I also attempted pulling down the latest prerelease module version separately and inserting it into the version constraint as a variable, which is also not supported. However, that's definitely out of scope for this issue.

### Proposal

Hashicorp developers previously shot down a [request for this](https://github.com/hashicorp/terraform/issues/31271#issuecomment-1160456206) in Terraform citing that "pre-release tags do not necessarily follow the normal semver order semantics" despite the fact that semver does have defined specs for prereleases. Additionally that "one usually does not desire any pre-releases to be included in a range of semantic versions" when there are a number of reasons for wanting to pull in the latest release or release range in a certain release channel.

My proposal is that OpenTofu support the full range of semver in version constraint strings on inexact operators i addition to the exact operators that are currently supported.

**Comment by cube2222** [comment]
Hey! Thanks for submitting the issue.

Could you expand on how those matches would work? Basically, you definitely don't want something like >= 1.6.0 to match 1.7.0-alpha1. Linking a relevant document that details this would be totally fine.

I suppose one approach would be to only match prereleases if the constraint contains a prerelease tag. An alternative would be to have a separate field which would indicate if prereleases should be ignored or not.

**Comment by Jellayy** [comment]
Yeah, I think ideally it would be best to have prerelease versions only be taken in when you specify.

Existing constraint strings such as `>= 1.6.0`, `~> 1.5.0` would be unaffected

But once you add a prerelease tag (`>=1.6.0-beta.1`), the hierarchy starts getting evaluated.

[semver](https://semver.org/) has an example hierarchy laid out:
```
1.0.0-alpha < 1.0.0-alpha.1 < 1.0.0-alpha.beta < 1.0.0-beta < 1.0.0-beta.2 < 1.0.0-beta.11 < 1.0.0-rc.1 < 1.0.0
```
How I've found this actually works on the their end is they sort alphabetically. So while `alpha` -> `beta` -> `rc` works, the spec will technically allow for any tags you add.

For example:
- `>= 1.0.0-rc.1` would take in `1.0.0-rc.1` then `1.0.0`
- `>= 1.0.0-beta.2` would take in `1.0.0-beta.2`, `1.0.0-beta11`, `1.0.0-rc.1`, then `1.0.0`

You could also possibly scope to one particular release channel with the `~>` operator, since it allows only the rightmost version component to increment:
- `~> 1.0.0-beta.1` would take in all `1.0.X-beta.X` releases
- `~> 1.0-rc.1` would take in all `1.X.X-rc.X` releases
- `~> 1-rc.1` could take in every release canidate?

Definitely open to see input from others on how this would work + any edge cases that might need this to be handled differently.

**Comment by apparentlymart** [comment]
One gotcha to watch out for here is that OpenTofu delegates to third-party libraries for much of the work related to version constraints. There are currently two different libraries in play due to code of different ages:

* [`github.com/hashicorp/go-version`](https://pkg.go.dev/github.com/hashicorp/go-version) is an older library used primarily for _module_ version selection, since that tends to need to be bug-compatible with very old versions of the product.
* [`github.com/apparentlymart/go-versions`](https://pkg.go.dev/github.com/apparentlymart/go-versions@v1.0.2/versions) is a newer library used for _provider_ version selection. It tends to produce better error messages for incorrect input, and originally it had introduced the special prerelease handling this issue is about, although [that was also later retrofitted into `hashicorp/go-version` too](https://github.com/hashicorp/go-version/commit/fbe76cfa08cd0fa2695a1e8feea4687221ed2b32).

---

The `apparentlymart/go-versions` incarnation of the rule this issue is discussing is implemented by the [`Set.WithoutUnrequestedPrereleases`](https://pkg.go.dev/github.com/apparentlymart/go-versions@v1.0.2/versions#Set.WithoutUnrequestedPrereleases) function, which conceptually subtracts all prerelease versions from the selection unless they are written out exactly. (Writing them out exactly is what "requested" means in the method name; refer to the docs of [`Set.Requests`](https://pkg.go.dev/github.com/apparentlymart/go-versions@v1.0.2/versions#Set.Requests) for more info on that.)

OpenTofu activates that behavior indirectly by calling [`versions.MeetingConstraints`](https://pkg.go.dev/github.com/apparentlymart/go-versions@v1.0.2/versions#MeetingConstraints), which is a higher-level helper function that builds a version set based on the result of parsing a constraints string. That function implicitly uses `WithoutUnrequestedPrereleases` because that's the most reasonable default behavior, but there's also [`versions.MeetingConstraintsExact`](https://pkg.go.dev/github.com/apparentlymart/go-versions@v1.0.2/versions#MeetingConstraintsExact) which includes only the constraints explicitly written and so _doesn't_ exclude the unrequested prereleases. Therefore the requested treatment could be achieved for provider version constraints by selectively using `MeetingConstraintsExact` in whatever situation we decide represents the user intent to include prereleases in the selection.

---

The `hashicorp/go-versions` library will probably make this behavior a little trickier to change, since the rule of excluding prereleases [is hard-coded in the constraint operator functions](https://github.com/hashicorp/go-version/blob/fcaa53267960208632e9308c93aa89535526a018/constraint.go#L235-L298). [`prereleaseCheck`](https://github.com/hashicorp/go-version/blob/fcaa53267960208632e9308c93aa89535526a018/constraint.go#L199-L217) takes two versions and returns `true` only if neither of them is a prerelease or they are both equal.

In principle we could take this opportunity to remove the remaining uses of `hashicorp/go-version`, but again I'd caution that `apparentlymart/go-versions` is probably not 100% bug-compatible with `hashicorp/go-version` and so switching could potentially cause unusual version constraints in existing modules to be interpreted differently than they are today.
